modules/tinymce/src/core/main/ts/fmt/FormatChanged.ts

```diff
@@ -5,101 +5,172 @@
 
-export type FormatCallbacks = Record<string, FormatChangeCallback[]>;
+export type RegisteredFormats = Record<string, CallbackGroup>;
 
-export interface FormatData {
-  similar?: boolean;
-  callbacks: FormatChangeCallback[];
+export interface UnbindFormatChanged {
+  unbind: () => void;
 }
 
-export type RegisteredFormats = Record<string, FormatData>;
+interface CallbackWithoutVars {
+  readonly state: Cell<boolean>;
+  readonly similar: boolean;
+  readonly callbacks: FormatChangeCallback[];
+}
 
-export interface UnbindFormatChanged {
-  unbind: () => void;
+interface CallbackWithVars {
+  readonly state: Cell<boolean>;
+  readonly similar: boolean;
+  readonly vars: FormatVars;
+  readonly callback: FormatChangeCallback;
 }
 
-const setup = (registeredFormatListeners: Cell<RegisteredFormats>, editor: Editor) => {
-  const currentFormats = Cell<Record<string, FormatChangeCallback[]>>({ });
+interface CallbackGroup {
+  // Batch all callbacks that don't have custom variables into one "similar=true" and one "similar=false" group
+  readonly withSimilar: CallbackWithoutVars;
+  readonly withoutSimilar: CallbackWithoutVars;
+  // For callbacks with custom variables, handle everything separately
+  readonly withVars: CallbackWithVars[];
+}
 
+const setup = (registeredFormatListeners: Cell<RegisteredFormats>, editor: Editor) => {
   registeredFormatListeners.set({});
 
   editor.on('NodeChange', (e) => {
-    updateAndFireChangeCallbacks(editor, e.element, currentFormats, registeredFormatListeners.get());
+    updateAndFireChangeCallbacks(editor, e.element, registeredFormatListeners.get());
   });
-};
 
-const updateAndFireChangeCallbacks = (editor: Editor, elm: Element, currentFormats: Cell<FormatCallbacks>, formatChangeData: RegisteredFormats) => {
-  const formatsList = Obj.keys(currentFormats.get());
-  const newFormats: FormatCallbacks = { };
-  const matchedFormats: FormatCallbacks = { };
-
-  // Ignore bogus nodes like the <a> tag created by moveStart()
-  const parents = Arr.filter(FormatUtils.getParents(editor.dom, elm), (node) => node.nodeType === 1 && !node.getAttribute('data-mce-bogus'));
-
-  // Check for new formats
-  Obj.each(formatChangeData, (data: FormatData, format: string) => {
-    Tools.each(parents, (node: Node) => {
-      if (editor.formatter.matchNode(node, format, {}, data.similar)) {
-        if (formatsList.indexOf(format) === -1) {
-          // Execute callbacks
-          Arr.each(data.callbacks, (callback: FormatChangeCallback) => {
-            callback(true, { node, format, parents });
-          });
-
-          newFormats[format] = data.callbacks;
-        }
+  editor.on('FormatApply FormatRemove', (e: EditorEvent<FormatEvent>) => {
+    const element = Optional.from(e.node)
+      .map((nodeOrRange) => Obj.get(nodeOrRange as RangeLikeObject, 'startContainer').getOr(nodeOrRange as Node))
+      .bind((node) => NodeType.isElement(node) ? Optional.some(node) : Optional.from(node.parentElement))
```

Comment:
This likely is going to be the wrong element in some cases, as `startContainer` could be a text node or an element itself. So if I'm understanding this logic correctly, we should only be using `parentElement` if it's a text node.